app/ui-message/client/messageBox/messageBoxAutogrow.js

```diff
@@ -0,0 +1,85 @@
+import _ from 'underscore';
+
+const replaceWhitespaces = (whitespaces: string) => `${ '&nbsp;'.repeat(whitespaces.length - 1) } `;
+
+export const setupAutogrow = (textarea: HTMLTextAreaElement, shadow: HTMLDivElement, callback: () => void) => {
+	if (textarea.updateAutogrow) {
+		return;
+	}
+
+	const width = textarea.clientWidth;
+	const height = textarea.clientHeight;
+	const { font, lineHeight } = window.getComputedStyle(textarea);
+
+	shadow.style.position = 'fixed';
+	shadow.style.top = '-10000px';
+	shadow.style.left = '-10000px';
+	shadow.style.width = `${ width }px`;
+	shadow.style.font = font;
+	shadow.style.lineHeight = lineHeight;
+	shadow.style.resize = 'none';
+	shadow.style.wordWrap = 'break-word';
+
+	const minHeight = height;
+	const maxHeight = parseInt(window.getComputedStyle(textarea).maxHeight, 10);
+
+	let lastWidth = width;
+	let lastHeight = minHeight;
+	let textLenght = 0;
+
+	const updateAutogrow = () => {
+		const { clientWidth: width, value: text } = textarea;
+
+		const isMaximumHeightReached = lastHeight >= maxHeight;
+		const isTextLengthOutdated = textLenght && textLenght < text.length;
+		const wasWidthChanged = width !== lastWidth;
+
+		if (isMaximumHeightReached && isTextLengthOutdated && !wasWidthChanged) {
+			return true;
+		}
+
+		const shadowText = text.replace(/</g, '&lt;')
+			.replace(/>/g, '&gt;')
+			.replace(/&/g, '&amp;')
+			.replace(/\n$/, '<br/>&nbsp;')
+			.replace(/\n/g, '<br/>')
+			.replace(/ {2,}/g, replaceWhitespaces);
+
+		if (wasWidthChanged) {
+			shadow.style.width = `${ width }px`;
+			lastWidth = width;
+		}
+
+		shadow.innerHTML = shadowText;
+
+		const shadowHeight = Math.max(shadow.clientHeight + 1, minHeight) + 1;
+		const height = Math.min(shadowHeight, maxHeight);
+
+		if (height === lastHeight) {
+			return true;
+		}
+
+		lastHeight = height;
+
+		const overflow = (height === maxHeight) ? 'hidden' : '';
+
+		if (height < maxHeight) {
+			textLenght = text.length;
+		}
+
+		textarea.style.overflow = overflow;
+		textarea.style.height = `${ height }px`;
+
+		callback && callback();
+	};
+
+	textarea.updateAutogrow = updateAutogrow;
+
+	const updateThrottled = _.throttle(updateAutogrow, 300);
+
+	const $textarea = $(textarea);
+	$textarea.on('change input', updateThrottled);
+	$textarea.on('focus', updateAutogrow);
+	$(window).resize(updateThrottled);
```

Comment:
should we provide a way to `off` this event?

Reply:
Ha! I knew there is at least one listener that could leak the `textarea`... I need to add another function to uninstall the autogrow and invoke it on `messageBox` template destruction.